packages/meditrak-server/src/routes/GETSurveyResponses/assertSurveyResponsePermissions.js

```diff
@@ -3,46 +3,78 @@
   }
   return true;
 };
 
-export const filterSurveyResponsesByPermissions = async (accessPolicy, surveyResponses, models) => {
-  if (accessPolicy.allowsSome(null, BES_ADMIN_PERMISSION_GROUP)) {
-    return surveyResponses;
+export const createSurveyResponseDBFilter = async (accessPolicy, models, criteria, options) => {
+  const dbConditions = { ...criteria };
+  const dbOptions = { ...options };
+
+  if (hasBESAdminAccess(accessPolicy)) {
+    return { dbConditions, dbOptions };
   }
-  const entities = await models.entity.findManyById(surveyResponses.map(sr => sr.entity_id));
-  const surveys = await models.survey.findManyById(surveyResponses.map(sr => sr.survey_id));
-  const permissionGroups = await models.permissionGroup.findManyById(
-    surveys.map(s => s.permission_group_id),
-  );
 
-  const entitiesById = keyBy(entities, 'id');
-  const surveysById = keyBy(surveys, 'id');
-  const permissionGroupsById = keyBy(permissionGroups, 'id');
+  const allPermissionGroupsNames = accessPolicy.getPermissionGroups();
+  const countryCodesByPermissionGroupId = {};
+
+  // Generate lists of country codes we have access to per permission group id
+  for (const permissionGroupName of allPermissionGroupsNames) {
+    const permissionGroup = await models.permissionGroup.findOne({ name: permissionGroupName });
+    if (permissionGroup) {
+      const countryNames = accessPolicy.getEntitiesAllowed(permissionGroupName);
+      countryCodesByPermissionGroupId[permissionGroup.id] = countryNames;
+    }
+  }
 
-  const filteredSurveyResponses = surveyResponses.filter(surveyResponse => {
-    const entity = entitiesById[surveyResponse.entity_id];
-    const survey = surveysById[surveyResponse.survey_id];
-    const permissionGroup = permissionGroupsById[survey.permission_group_id];
+  // Join SQL table with entity and survey tables
+  // Running the permissions filtering is much faster with joins
+  dbOptions.multiJoin = [
+    {
+      joinWith: TYPES.SURVEY,
+      joinCondition: [`${TYPES.SURVEY}.id`, `${TYPES.SURVEY_RESPONSE}.survey_id`],
+    },
+    {
+      joinWith: TYPES.ENTITY,
+      joinCondition: [`${TYPES.ENTITY}.id`, `${TYPES.SURVEY_RESPONSE}.entity_id`],
+    },
+  ];
+
+  // If columns weren't specified, avoid returning the joined columns
+  if (!dbOptions.columns) {
+    dbOptions.columns = ['survey_response.*'];
+  }
 
-    return accessPolicy.allows(entity.country_code, permissionGroup.name);
-  });
+  // Check the country code of the entity exists in our list for the permission group
+  // of the survey
+  dbConditions[RAW] = {
+    sql: `
+    (
+      ARRAY[entity.country_code]::TEXT[]
+      <@
+      ARRAY(
+        SELECT TRIM('"' FROM JSON_ARRAY_ELEMENTS(?::JSON->survey.permission_group_id)::TEXT)
+      )
+    )`,
+    parameters: JSON.stringify(countryCodesByPermissionGroupId),
+  };
 
-  return filteredSurveyResponses;
+  return { dbConditions, dbOptions };
 };
```

Comment:
Same comments in this file as the previous (maybe reflecting the opportunity to make some bits more DRY, but I'm also not worried at this stage in the game)